diff --git a/build_tools/bazel/build_defs.oss.bzl b/build_tools/bazel/build_defs.oss.bzl
--- a/build_tools/bazel/build_defs.oss.bzl
+++ b/build_tools/bazel/build_defs.oss.bzl
@@ -4,24 +4,35 @@
 # See https://llvm.org/LICENSE.txt for license information.
 # SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
-load("@llvm-project//mlir:tblgen.bzl", "gentbl_cc_library", "gentbl_filegroup", "td_library")
+# load("@llvm-project//mlir:tblgen.bzl", "gentbl_cc_library", "gentbl_filegroup", "td_library")
 
 # All load statements must come first in Starlark.
 load("@rules_cc//cc:defs.bzl", "cc_binary", "cc_library", "cc_test")
-load("@rules_shell//shell:sh_test.bzl", "sh_test")
-load(
-    "//build_tools/bazel:iree_cc_fuzz.bzl",
-    _iree_cc_fuzz = "iree_cc_fuzz",
-    _iree_compiler_cc_fuzz = "iree_compiler_cc_fuzz",
-    _iree_runtime_cc_fuzz = "iree_runtime_cc_fuzz",
-)
+# load("@rules_shell//shell:sh_test.bzl", "sh_test")
+# load(
+#     "//build_tools/bazel:iree_cc_fuzz.bzl",
+#     _iree_cc_fuzz = "iree_cc_fuzz",
+#     _iree_compiler_cc_fuzz = "iree_compiler_cc_fuzz",
+#     _iree_runtime_cc_fuzz = "iree_runtime_cc_fuzz",
+# )
 
 """Common Bazel definitions for IREE."""
 
 # Re-export fuzz rules for external use.
-iree_cc_fuzz = _iree_cc_fuzz
-iree_compiler_cc_fuzz = _iree_compiler_cc_fuzz
-iree_runtime_cc_fuzz = _iree_runtime_cc_fuzz
+# iree_cc_fuzz = _iree_cc_fuzz
+# iree_compiler_cc_fuzz = _iree_compiler_cc_fuzz
+# iree_runtime_cc_fuzz = _iree_runtime_cc_fuzz
+
+def generic_pass(**kwargs):
+    pass
+
+gentbl_cc_library = generic_pass
+gentbl_filegroup = generic_pass
+td_library = generic_pass
+sh_test = generic_pass
+iree_runtime_cc_fuzz = generic_pass
+iree_compiler_cc_fuzz = generic_pass
+iree_cc_fuzz = generic_pass
 
 def defaulting_select(selector):
     """Pass through to select() with special semantics when converting to CMake.
diff --git a/runtime/src/iree/base/tracing/BUILD.bazel b/runtime/src/iree/base/tracing/BUILD.bazel
--- a/runtime/src/iree/base/tracing/BUILD.bazel
+++ b/runtime/src/iree/base/tracing/BUILD.bazel
@@ -26 +25,0 @@
-        "tracy",
@@ -37,7 +35,0 @@
-config_setting(
-    name = "_tracy_enable",
-    flag_values = {
-        ":tracing_provider": "tracy",
-    },
-)
-
@@ -52 +43,0 @@
-        ":_tracy_enable": ":tracy",
@@ -74,18 +64,0 @@
-
-#===------------------------------------------------------------------------===#
-# Tracy
-#===------------------------------------------------------------------------===#
-
-iree_runtime_cc_library(
-    name = "tracy",
-    srcs = ["tracy.cc"],
-    hdrs = ["tracy.h"],
-    defines = [
-        "IREE_TRACING_PROVIDER_H=\\\"iree/base/tracing/tracy.h\\\"",
-        "IREE_TRACING_MODE=2",
-    ],
-    deps = [
-        "//runtime/src/iree/base:core_headers",
-        "@tracy_client//:runtime",
-    ],
-)
