diff --git a/build_tools/bazel/build_defs.oss.bzl b/build_tools/bazel/build_defs.oss.bzl
index 090c87f860..ed1ab94fd0 100644
--- a/build_tools/bazel/build_defs.oss.bzl
+++ b/build_tools/bazel/build_defs.oss.bzl
@@ -4,23 +4,33 @@
 # See https://llvm.org/LICENSE.txt for license information.
 # SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
-load("@llvm-project//mlir:tblgen.bzl", "gentbl_cc_library", "gentbl_filegroup", "td_library")
+# load("@llvm-project//mlir:tblgen.bzl", "gentbl_cc_library", "gentbl_filegroup", "td_library")
 
 # All load statements must come first in Starlark.
 load("@rules_cc//cc:defs.bzl", "cc_binary", "cc_library", "cc_test")
-load(
-    "//build_tools/bazel:iree_cc_fuzz.bzl",
-    _iree_cc_fuzz = "iree_cc_fuzz",
-    _iree_compiler_cc_fuzz = "iree_compiler_cc_fuzz",
-    _iree_runtime_cc_fuzz = "iree_runtime_cc_fuzz",
-)
+# load(
+#     "//build_tools/bazel:iree_cc_fuzz.bzl",
+#     _iree_cc_fuzz = "iree_cc_fuzz",
+#     _iree_compiler_cc_fuzz = "iree_compiler_cc_fuzz",
+#     _iree_runtime_cc_fuzz = "iree_runtime_cc_fuzz",
+# )
 
 """Common Bazel definitions for IREE."""
 
 # Re-export fuzz rules for external use.
-iree_cc_fuzz = _iree_cc_fuzz
-iree_compiler_cc_fuzz = _iree_compiler_cc_fuzz
-iree_runtime_cc_fuzz = _iree_runtime_cc_fuzz
+# iree_cc_fuzz = _iree_cc_fuzz
+# iree_compiler_cc_fuzz = _iree_compiler_cc_fuzz
+# iree_runtime_cc_fuzz = _iree_runtime_cc_fuzz
+
+def generic_pass(**kwargs):
+    pass
+
+gentbl_cc_library = generic_pass
+gentbl_filegroup = generic_pass
+td_library = generic_pass
+iree_runtime_cc_fuzz = generic_pass
+iree_compiler_cc_fuzz = generic_pass
+iree_cc_fuzz = generic_pass
 
 def defaulting_select(selector):
     """Pass through to select() with special semantics when converting to CMake.
