diff --git a/apt/private/version_constraint.bzl b/apt/private/version_constraint.bzl
index 531fa01..b3a3d30 100644
--- a/apt/private/version_constraint.bzl
+++ b/apt/private/version_constraint.bzl
@@ -2,11 +2,19 @@
 
 load(":version.bzl", version_lib = "version")
 
+_VERSION_CONSTRAINT_OPERATORS = ["<<", ">>", "<=", ">=", "="]
+
 def _parse_version_constraint(rawv):
-    vconst_i = rawv.find(" ")
-    if vconst_i == -1:
-        fail('invalid version string %s expected a version constraint ">=", "=", ">=", "<<", ">>"' % rawv)
-    return (rawv[:vconst_i], rawv[vconst_i + 1:])
+    rawv = rawv.strip()
+    for op in _VERSION_CONSTRAINT_OPERATORS:
+        if rawv.startswith(op):
+            version = rawv[len(op):].strip()
+            if version:
+                return (op, version)
+
+            break
+
+    fail('invalid version string %s expected a version constraint ">=", "=", "<=", "<<", ">>"' % rawv)
 
 def _parse_dep(raw):
     raw = raw.strip()  # remove leading & trailing whitespace
