--- xla/tsl/profiler/convert/BUILD
+++ xla/tsl/profiler/convert/BUILD
@@ -104,9 +104,7 @@ cc_library(
     srcs = ["trace_events_to_json.cc"],
     hdrs = ["trace_events_to_json.h"],
     copts = tf_profiler_copts(),
-    visibility = internal_visibility([
-        "//xla/tsl/profiler:internal",
-    ]),
+    visibility = ["//visibility:public"],
     deps = [
         ":trace_container",
         "//xla/tsl/platform:types",
@@ -150,9 +148,7 @@ cc_library(
     srcs = ["xplane_to_trace_events.cc"],
     hdrs = ["xplane_to_trace_events.h"],
     copts = tf_profiler_copts(),
-    visibility = internal_visibility([
-        "//xla/tsl/profiler:internal",
-    ]),
+    visibility = ["//visibility:public"],
     deps = [
         ":trace_container",
         "//xla/tsl/platform:types",
